### ISSUE TITLE
Incremental update / sync queue leads to OOM Killer on RPi

### ISSUE THREAD

**Issue by macearl** [jellyfin/jellyfin-kodi#472]
**Describe the bug**
When there are a lot of updates the incremental update leads to Kodi being killed by the OOM Killer on Devices with a low amount of RAM (e.g. RPi 3 1GB)

**To Reproduce**
1. Change a lot of entries in your jellyfin installation (e.g. have the brilliant idea to move all metadata to a different drive, for me it shows ~2200 updates) 
2. Confirm the question whether or not you want to continue with the incremental update 
3. Just wait until it crashes (for me it gets up to processing ~800 items) 

**Expected behavior**
The sync should complete without crashes 

**Logs**
```
Feb 20 16:31:59 RPi3Bplus kernel: Out of memory: Kill process 358 (kodi.bin) score 859 or sacrifice child
Feb 20 16:31:59 RPi3Bplus kernel: Killed process 358 (kodi.bin) total-vm:1304528kB, anon-rss:657496kB, file-rss:0kB, shmem-rss:0kB
Feb 20 16:31:59 RPi3Bplus kernel: oom_reaper: reaped process 358 (kodi.bin), now anon-rss:0kB, file-rss:0kB, shmem-rss:0kB
Feb 20 16:31:59 RPi3Bplus kernel: vchiq: 0: ignoring ERROR from callback to service 5004
```

Enabling debug logging in the addon leads to the process taking roughly 18 hours as it only increases the amount of items to be processed very slowly. Unfortunately after letting it continue for 18 hours I forgot to save the log file... 

**System (please complete the following information):**
 - OS: Libreelec 9.2.6 
 - Jellyfin Version: 16.4
 - Kodi Version: 18.9
 - Addon Version: 0.7
 - Playback Mode: Addon

**Additional context**
I suspect its a similar issue to https://github.com/jellyfin/jellyfin-kodi/issues/350

The incremental update probably uses different download code or loads every entry to be updated into a queue first which leads to high memory usage?

I haven't been able to debug this further yet, any pointers to the responsible code would be appreciated.

**Comment by mcarlton00** [comment]
The incremental sync is a bit of a mess that I haven't gotten to yet, unfortunately.  It starts here: https://github.com/jellyfin/jellyfin-kodi/blob/ce111cfb61a1443d5151959ba18d4ffae5768cd0/jellyfin_kodi/library.py#L398.  It then sorts the output to one of 3 queue, depending on what change has occurred (item removed, item added/modified, or a userdata update).  These queues are then worked on by the various `worker_XXX` functions.  That's about as far as I can remember from the last time I went through it.

Some gut feeling stuff:

- i doubt it's the 3 queues themselves getting too large.  The Kodi Sync Queue just returns a list of item IDs, so those queues themselves should be fairly small, all things considered.  This also makes it annoying since it's only item IDs, and we need to make another api call for every item in the list to figure out what it is.
- My initial thought is that it's in the `worker_XXX` functions.  From a quick glance they don't seem to be getting limited by the max threads, so it's probably trying to spin up too many at once

**Comment by macearl** [comment]
Finally had some time yesterday and played around with a couple of changes, i probably missed something obvious but here goes:

I tried introducing `thread.join()` and semaphores in pretty much all the worker functions to reduce the amount of parallel threads (either in one, some or all of them at the same time).

As far as I can tell nothing changed at all, as soon as the number of items to be processed increases and reaches around 400-450 (I played around on another system running xbian which seems to be quite bit more of a memory hog than libreelec) it crashes.

So maybe the problem does begin even before the updates are processed? I haven't really used Web pdb before and haven't yet found a way to actually inspect variables closer.

It looks like I have also killed my xbian install through all the crashes, I'll have to take a look at the fs now and see if I can restore that, if not I'll have to set up another jellyfin docker instance or something in which I can generate new updates without infuriating my family members again, as my strategy to restore to a state which has a lot of updates was "just restore a btrfs snapshot"

**Comment by macearl** [comment]
OK, so I've now set up a separate docker installation, added all my media sources and generated 1500 updates by redownloading some metadata.
I made some changes to the plugin (included the update worker processes into the download limit and not just the currently running downloads) and kodi finished the sync queue without a crash. Great! Just as a quick sanity check I restored my backup and ran another sync without any changes to the code with the same 1500 updates and it also finished without a crash... well, that was unexpected... 

Since my last test I did update my main server to jellyfin 10.7 (deb package) and I've also pulled the 10.7 docker image. So I guess something changed with jellyfin 10.7 which seems to have fixed the issue? (perhaps the api delivers less data?) 

Now to truly be sure I would need to generate a lot of updates on my main server again, but as that could result in another round of family tech support I think I'm just gonna keep my fingers crossed that this is fixed and just in case keep an eye open if I do run into another situation where my main server does have a lot of updates. 

Not sure if you want to keep this issue open to see if someone else has similar issues, but I think it's safe to close it right now, if the issue arises again it can then be reopened
